### PR TITLE
Add appveyor

### PR DESCRIPTION
**Pull request by ProgramMax**

### CODE REVIEW

File: test/example.c [hunk 1]
```diff
@@ -575,25 +684,55 @@ int main(argc, argv)
     (void)argc;
     (void)argv;
 #else
-    test_compress(compr, comprLen, uncompr, uncomprLen);
+    if (argc > 1) {
+        if (strcmp(argv[1], "--appveyor") == 0) {
```
Comment (dankegel) [inline test/example.c]: Odd to hardcode name of CI tool and its output path. 

Suggestion:
 --junit=path-to-output-file might be more appropriate.  
Or whatever googletest does.

File: test/example.c [hunk 2]
```diff
@@ -575,25 +684,55 @@ int main(argc, argv)
...
-    test_gzio((argc > 1 ? argv[1] : TESTFILE),
-              uncompr, uncomprLen);
+            output = fopen("tooling/appveyor/junit-results.xml", "w+");
+            if (!output) {
+                fprintf(stderr, "Could not open junit file");
+                exit(1);
+            }
+            fprintf(output, "<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n");
+            fprintf(output, "<testsuites>\n");
+            fprintf(output, "\t<testsuite name=\"zlip example suite\">\n");
```
Comment (dankegel) [inline test/example.c]: zlip ?

File: test/example.c [hunk 3]
```diff
@@ -19,13 +19,76 @@
...
+            fprintf(output, "\n\t\t\t<failure>%s error: %d</failure>\n\t\t", result.message, result.err); \
+		} else { \
+            fprintf(stderr, "%s error: %d\n", result.message, result.err); \
+            exit(1); \
+		} \
+    } else if (result.result == FAILED_WITHOUT_ERROR_CODE) { \
+        if (is_junit_output) { \
+            fprintf(output, "\n\t\t\t<failure>%s</failure>\n\t\t", result.message); \
+		} else { \
+            fprintf(stderr, result.message); \
```
Comment (dankegel) [inline test/example.c]: Might want to use "%s", result.message:

../test/example.c:707:5: warning: format not a string literal and no format arguments [-Wformat-security]
../test/example.c:711:5: warning: format not a string literal and no format arguments [-Wformat-security]
     HANDLE_TEST_RESULTS(output, result, "deflate", is_junit_output);
     ^~~~~~~~~~~~~~~~~~~
../test/example.c:711:5: warning: format not a string literal and no format arguments [-Wformat-security]
../test/example.c:713:5: warning: format not a string literal and no format arguments [-Wformat-security]
     HANDLE_TEST_RESULTS(output, result, "inflate", is_junit_output);
     ^~~~~~~~~~~~~~~~~~~
../test/example.c:713:5: warning: format not a string literal and no format arguments [-Wformat-security]
../test/example.c:716:5: warning: format not a string literal and no format arguments [-Wformat-security]
     HANDLE_TEST_RESULTS(output, result, "large deflate", is_junit_output);
     ^~~~~~~~~~~~~~~~~~~
../test/example.c:716:5: warning: format not a string literal and no format arguments [-Wformat-security]
../test/example.c:718:5: warning: format not a string literal and no format arguments [-Wformat-security]
     HANDLE_TEST_RESULTS(output, result, "large inflate", is_junit_output);

File: test/example.c [hunk 4]
```diff
@@ -685,7 +685,7 @@ int main(argc, argv)
     (void)argv;
 #else
     if (argc > 1) {
-        if (strcmp(argv[1], "--appveyor") == 0) {
+        if (strcmp(argv[1], "--junit") == 0) {
```
Comment (dankegel) [inline test/example.c:701]: Nice.  But the hardcoded path is still a problem; if you run from the build directory (as out of tree builds tend to do on Linux), it fails to find the output file's directory:

```1: Test command: /home/dank/src/zlib/btmp/example "--appveyor"
1: Test timeout computed to be: 10000000
1: Could not open junit filezlib version 1.2.11.1-motley = 0x12b1, compile flags = 0xa9
```
And if you do an in-tree build using the Makefile generator, it clobbers Makefile :-(

File: test/example.c [hunk 5]
```diff
@@ -19,13 +19,76 @@
...
+#define EXIT_ON_ERROR(err, msg) { \
     if (err != Z_OK) { \
         fprintf(stderr, "%s error: %d\n", msg, err); \
         exit(1); \
     } \
 }
 
+#define RETURN_ON_ERROR_WITH_MESSAGE(err, msg, result) { \
+    if (err != Z_OK) { \
+        result.err = err; \
```
Comment (dankegel) [inline test/example.c]: delicious clash if actual parameter is not named err?
```
test/example.c: In function ‘test_inflate’:
test/example.c:337:34: error: expected identifier before numeric constant
  337 |     RETURN_ON_ERROR_WITH_MESSAGE(7777, "inflateInit", result);
      |                                  ^~~~
test/example.c:45:16: note: in definition of macro ‘RETURN_ON_ERROR_WITH_MESSAGE’
   45 |         result.err = err; \
      |                ^~~
```

**Comment by dankegel** [comment]
If you're going to remove the exit(1) from the error check, you need to add it at the end of main(), else ctest will say the test succeeded!

File: test/example.c (hunk 4 above)
Comment (dankegel) [reply test/example.c:701]: If you add --junit to the tests in CMakeLists.txt, both example and example64 output to the same file, and the results of one are overwritten.  Hrm.

**Comment by ProgramMax** [comment]
CMake shouldn't be passing --junit, so it can preserve the old stdout prints.
I tried to make this change a noop outside of the CI.

As a result, it will continue hitting the exit(1)s I believe, right?

**Comment by ProgramMax** [comment]
Finally got it. You were right about returning non-zero in case of error.

Okay, so main() SHOULD return non-zero if there is a critical error.
(A test could fail but be non-critical and thus people might want the whole to pass.)
I assumed if a test failed the whole thing would be marked red. But that's not the case. Returning 0/non-0 determines green/red.

I don't want to exit(1) on error because then the other tests don't run. So I'll keep track of whether or not any test failed and return 1 if so.

**Comment by ProgramMax** [comment]
@dankegel I think this is ready for a real review if you would like to take a look.

I moved the file & line number from the testcase into the failure tags.
I think AppVeyor doesn't recognize it, since it isn't part of the JUnit XSL. It makes more sense to be on the failure tag, if that other CI supports it being there.

**Pull request merged**